SI-7331 tb.parse returns unpositioned trees

densh opened the pull request:
This pull request gets rid off code wrapping that was previously used by toolbox to get into correct parsing mode. Instead combination of templateStats/accept(EOF) is used. This is the same solution as the one used in repl and built-in scriptRunner

This pull request doesn't attempt to generalize this approach in any way and re-use it all over the place due to the caution of accidental compatibility breakage. I plan to do it separately against master.

Additionally there are a few more changes that make importers be aware of positions and a test for that (via @jedesah).

densh committed:
SI-7331 tb.parse returns unpositioned trees

This commit gets rid off code wrapping that was previously used by
toolbox to get into correct parsing mode. Instead combination of
templateStats/accept(EOF) is used. This is the same solution as the one
used in repl and built-in scriptRunner

This pull request doesn't attempt to generalize this approach in any
way and re-use it all over the place due to the caution of possible
accidental compatibility breakage. I plan to do it separately against
master.

Additionally there are a few more changes that make importers be aware
of positions and a test for that (via @jedesah).

densh commented:
@jedesah it'd be great if you tried this version of the fix on your code base. I don't expect anything to break though.

retronym commented:
Could you add a test with a missing brace? I remember the REPL code you showed me yesterday only called `accept(EOF)` if there had been no errors.

densh commented:
I've just updated the pr with suggested extra test. Toolbox fails on the first encountered error so this should not be an issue. I will investigate necessity of that check during my refactoring against master.

retronym commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -968,7 +968,7 @@ trait Trees extends api.Trees { self: SymbolTable =>
     override def isEmpty = true
     super.setPos(NoPosition)
-    override def setPos(pos: Position) = { assert(false); this }
+    override def setPos(pos: Position) = this
```
This warrants a comment to explain why it is safe to discard the position (any why it would be unsafe to use it on a shared tree like this.)

How does this compare to `EmptyTree`?, That extends:

``` scala
trait CannotHaveAttrs extends Tree {
    override def canHaveAttrs = false

    private def unsupported(what: String, args: Any*) =
      throw new UnsupportedOperationException(s"$what($args) inapplicable for "+self.toString)

    super.setPos(NoPosition)
    override def setPos(pos: Position) = unsupported("setPos", pos)
```

... which throws an error in `setPos`. `EmptyTrees` are used, e.g. to encode unspecified upper/lower bounds for type parameters.

densh replied:
This must have been added in previous pr to correct possible errors during tree import. I guess I'll undo this change and add extra checks to tree importer instead.

densh commented:
I've just updated pr without `setPos` change on `emptyValDef` and added extra test that checks that NoPosition trees like `EmptyTree` and `emptyValDef` are imported correctly.

xeno-by commented:
Why was the setPos change reverted? Have you followed the history of this change? (https://issues.scala-lang.org/browse/SI-7331)

xeno-by commented on src/reflect/scala/reflect/internal/Importers.scala:
```diff
@@ -443,7 +443,10 @@ trait Importers extends api.Importers { self: SymbolTable =>
       })
       tryFixup()
-      mytree
+      if(tree.pos != NoPosition)
```
Whitespace.

densh commented:
Current `setPos` behaviour is consistent with master even though it fails with different exception (as pointed out by @retronym ). I believe that @jedesah did this change to prevent possible failures during position import. In order to prevent this failure i tweaked import logic instead (i.e. don't set position if it's NoPosition).

xeno-by commented:
Have you taken a look at the original pull request that's linked in the issue's comment?

densh commented:
In previous pr @paulp talks about inconsistencies in CannotHaveAttrs. To fix it we need to backport it from master first. Won't it break binary compatibility in 2.10?

jedesah commented:
Relative to the removal or not of the assert in the setPos function, I don't have a strong opinion on the matter. It made sense to me after reading @paulp 's comments on the original pull request.

I ran the test suite of my library and it's green across the board. Yay!

Thx a lot @densh for jumping in and providing a better solution to the problem and coming up with an alternative PR in a very timely manner. It's very much appreciated.

densh commented:
/cc @retronym

retronym commented on src/reflect/scala/reflect/internal/Importers.scala:
```diff
@@ -443,7 +443,10 @@ trait Importers extends api.Importers { self: SymbolTable =>
       })
       tryFixup()
-      mytree
+      if (tree.pos != NoPosition)
```
Needs a comment referring back to `emptyValDef` / `EmptyTree` and etc.

retronym commented:
I think this is the right patch for 2.10.x. Just add that comment and it LG to me.

densh commented:
Just updated pr with explanatory comment as suggested by @retronym

retronym commented:
LGTM
